### ISSUE TITLE
Difference between Embedding formulations

### ISSUE THREAD

**Issue by CDitzel** [lucidrains/vit-pytorch#60]
Hi, me again. Skimming through your repositories, I am wondering, what is the difference between using embeddings like

nn.Embedding(num_tokens, dim)

vs. 

nn.Linear(num_tokens, dim)

vs. 

 nn.Parameter(torch.randn(num_tokens, dim))

I have seen all three in code of yours. All of them are trainable, so is it just a matter of style?

**Comment by lucidrains** [comment]
@CDitzel haha yup, they are actually all the same underneath, just a trainable matrix (linear comes with an extra bias tensor). however, they have wrappers that allow them to behave differently when you give it an input

1. `nn.Embedding` selects the rows of the given matrix, given a list of integers

2. `nn.Linear` does the einsum operation `...d, d e -> ...e`

3. `nn.Parameter` basically just makes a tensor trainable (receive gradients and updates on step). this is the lowest level you can go, so actually, you can define your entire deep neural network with just nn.Parameters and manually do all the above

Hope this helps!

**Comment by CDitzel** [comment]
thank you very much. so my basic assumption was correct then and it depends on the case at hand. 

For ViT, nn.Embedding is not appropriate unless the index of the image patch would be used for get the corresponding row. so it is easier in this case to just use self.patch_to_embedding = nn.Linear(patch_dim, dim)

**Comment by lucidrains** [comment]
@CDitzel Yup! the authors of ViT have basically proved to us that attention doesn't need to work with discrete number of embeddings

**Comment by CDitzel** [comment]
another important aspect is the fact that nn.Embedding used e.g. in NLP causes each token to have its own individual embedding vector of arbitrary length whereas using e.g. nn.Parameter on image patches shares its parameters across every token. Imho this should in theory be less effective than the first variant, should it not?
